enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,532 @@
+
+By using this feature gate with `TechPreviewNoUpgrade` `FeatureSet`, users will prevent their clusters from upgrading to newer versions of OpenShift as with all other feature gates.
+
+The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.
+If the feature gate  is listed (via a custom or technical preview feature set), the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for any platform.
+
+Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the `TechPreviewNoUpgrade` feature gate and will automatically upgrade as part of a 4.N to 4.N+1 upgrade using default feature set.
+
+Once all providers are moved out-of-tree, the feature gate will be safe to remove.
+
+This feature gate will be commonly used in CI tasks, to ensure upgrade stability. It will also be used in conjunction with `CSIMigration*` feature gates, to start volume migration during upgrades. The implementation will require changes in [installer](https://github.com/openshift/installer) to support feature gates at install time.
+
+#### Bootstrap changes
+
+Once an out-of-tree provider is released (GA), the `CCM` will be created as a static pod on the bootstrap node, to ensure swift removal of the `node.cloudprovider.kubernetes.io/uninitialized: true` taint from any newly created `Nodes`. Later stages, including cluster upgrades will be managed by an operator, which will ensure stability of the configuration, and will run the `CCM` in a `Deployment`.
+Initial usage of the static pod is justified by the need for `CCM` to initialise `Nodes` before the `kube-scheduler` is able to schedule regular worklods (eg the operator for `CCM`).
+
+The order will be the following:
+
+1) Start `etcd` and `kube-api-server`.
+2) Run the `CCM`, which will be provisioned as a static pod.
+3) Start the `kube-scheduler`.
+
+`CCCMO` will provide a way to generate initial manifests that allow deployment of `CCM` as a static pod on the bootstrap machine.
+This will be used with the `bootkube.sh` [script](https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template).
+
+At the initial stage of `CCM` installation, the installer creates a `ConfigMap` with a `cloud.conf` key that contains configuration of `CCM`.
+The contents of `cloud.conf` are static and generated by the installer:
+
+Example for OpenStack:
+
+```txt
+[Global]
+secret-name = openstack-credentials
+secret-namespace = kube-system
+```
+
+Resources deployed here will be destroyed by the bootstrap cleanup process.
+
```

Comment:
I miss something about the feature gate topic here in the section. The installer allows to set them? How does the bootstrap phase know? There is no feature gate CR yet.

Reply:
This section is describing the post GA state, when that happens the conditional blocking on the feature gate will be removed and components will deploy as out of tree during bootstrap. The top of this section has been updated to include a note that this is for GA, I think this is resolved